garage/envs/base.py

```diff
@@ -1,17 +1,156 @@
+from garage.envs.env_spec import EnvSpec
+from garage.misc.overrides import overrides
+from garage.spaces import Box
+from garage.spaces import Discrete
+from garage.spaces import Product
+
+
+class GarageEnv(gym.Wrapper, Parameterized, Serializable):
+    """
+    Returns a Garage wrapper class for gym.Env.
+
+    In order to provide pickling (serialization) and parameterization
+    for gym.Envs, they must be wrapped with GarageEnv. This ensures
+    compatibility with existing samplers and checkpointing when the
+    envs are passed internally around garage.
+
+    Furthermore, GarageEnv silently converts action_space and
+    observation_space from gym.Spaces to garage.spaces.
+
+    Args:
+        env (gym.Env): the env that will be wrapped
+    """
+
+    def __init__(self, env):
+        Serializable.quick_init(self, locals())
+        Parameterized.__init__(self)
+        super(GarageEnv, self).__init__(env)
+
+    @overrides
+    def action_space(self):
+        """
+        Classes inheriting from GarageEnv need to convert
+        action_space from gym.space to garage.space.
+
+        Returns:
+            NotImplementedError
+        """
+        raise NotImplementedError
+
+    @overrides
+    def observation_space(self):
+        """
+        Classes inheriting from GarageEnv need to convert
+        observation_space from gym.space to garage.space.
+
+        Returns:
+            NotImplementedError
+        """
+        raise NotImplementedError
+
+    def close(self):
+        """
+        Close the wrapped env.
+
+        Returns:
+            None
+        """
+        self.env.close()
+
+    def get_params_internal(self, **tags):
+        """
+        Returns an empty list if env.get_params() is called.
+
+        Returns:
+            An empty list
+        """
+        return []
+
+    @property
+    def horizon(self):
+        """
+        Get the maximum episode steps for the wrapped env.
+
+        Returns:
+            max_episode_steps (int)
+        """
+        if self.env.spec is not None:
+            return self.env.spec.max_episode_steps
+        else:
+            return NotImplementedError
+
+    def log_diagnostics(self, paths, *args, **kwargs):
+        pass
+
+    def render(self, *args, **kwargs):
+        return self.env.render(*args, **kwargs)
+
+    def reset(self, **kwargs):
+        return self.env.reset(**kwargs)
+
+    @property
+    def spec(self):
+        """
+        Returns an EnvSpec.
+
+        Returns:
+            spec (garage.envs.EnvSpec)
+        """
+        return EnvSpec(
+            observation_space=self.observation_space,
+            action_space=self.action_space,
+        )
+
+    def step(self, action):
+        return self.env.step(action)
+
+    def wrapped_env(self):
+        """
+        Returns the original env wrapped by this class.
+
+        Returns:
+            env (gym.Env)
+        """
+        return self.env
+
+    def _to_garage_space(self, space):
```

Comment:
is this ever called anywhere?

Reply:
You could make this also abstract (e.g. NotImplementedError) and then implement it concretely in TfEnv and TheanoEnv